rfcs/20231005-hybrid-quantization.md

```diff
@@ -0,0 +1,158 @@
+output_batch_dimension, output_feature_dimension, output_spatial_dimensions,
+feature_group_count, batch_group_count, precision_config), lhs, rhs)`.
+
+#### Current constraints
+
+* If the operation uses quantized tensors:
+  * (C28) `is_quantized_tensor(lhs) and is_quantized_tensor(rhs) and
+    is_quantized_tensor(result)`.
+  * (C29) `storage_type(lhs) =  storage_type(rhs)`.
+  * (C30) `expressed_type(lhs) = expressed_type(rhs) = expressed_type(result)`.
+  * (C31) If `is_per_tensor_quantized(rhs)`,
+    then `is_per_tensor_quantized(result)`.
+  * (C32) If `is_per_axis_quantized(rhs)`, then
+    `quantization_dimension(rhs) = kernel_output_feature_dimension`.
+  * (C33) If `is_per_axis_quantized(result)`, then
+    `quantization_dimension(result) = output_feature_dimension`.
+
+#### Proposed constraints
+
+* If the operation uses quantized tensors:
+  * (C28) `is_quantized(lhs) = is_quantized(result) and is_quantized(rhs)`.
+  * (C29) If `is_per_axis_quantized(rhs)`,
+    then `quantization_dimension(rhs) = kernel_output_feature_dimension`.
+  * (C30) If `is_per_axis_quantized(result)`, then 
+    `quantization_dimension(result) = output_feature_dimension`.
+  * If `is_quantized(lhs)`:
+    * (C31) `storage_type(lhs) = storage_type(rhs)`.
+    * (C32) `expressed_type(lhs) = expressed_type(rhs) = expressed_type(result)`.
+    * (C33) If `is_per_tensor_quantized(rhs)`, then
+      `is_per_tensor_quantized(result)`.
+  * If `!is_quantized(lhs)`:
+    * (C34) `element_type(lhs) = expressed_type(rhs) = element_type(result)`.
+
+### dot_general
+
+#### Operation semantics for hybrid op
+
+For hybrid quantized types, performs `hybrid_dequantize_then_op( lambda lhs,
+rhs: dot_general(lhs, rhs, lhs_batching_dimensions, rhs_batching_dimensions,
+lhs_contracting_dimensions, rhs_contracting_dimensions, precision_config), lhs,
+rhs)`.
+
+#### Current constraints
+
+* If the operation uses quantized tensors:
+  * (C14) `is_quantized(lhs) and is_quantized(rhs) and is_quantized(result)`.
+  * (C15) `storage_type(lhs) = storage_type(rhs)`.
+  * (C16) `expressed_type(lhs) = expressed_type(rhs) = expressed_type(result)`.
+  * (C17) `zero_points(rhs) = 0`.
+
+#### Proposed constraints
+
+* If the operation uses quantized tensors:
+  * (C14) `is_quantized(lhs) = is_quantized(result) and is_quantized(rhs)`.
+  * (C15) `zero_points(rhs) = 0`.
+  * If `is_quantized(lhs)`:
+    * (C16) `storage_type(lhs) = storage_type(rhs)`.
+    * (C17) `expressed_type(lhs) = expressed_type(rhs) = expressed_type(result)`.
+  * If `!is_quantized(lhs)`:
+    * (C18) `element_type(lhs) = expressed_type(rhs) = element_type(result)`.
+
+## Other related topics not covered in this RFC
+
+### Dynamic Range Quantization
+
+Dynamic range quantization(DRQ) is a different quantization scheme, which has
+the same type signature as weight-only quantization. A dynamic range quantized
+graph also accepts input in float and weight in quantized type. Instead of
```

Comment:
Dynamic range quantization(DRQ) is a different quantization scheme, which is represented in comparable MLIR dialects using the same type signature as this proposed weight-only quantization representation.

Reply:
Done.